### PR TITLE
Feat/7725 url join

### PR DESCRIPTION
**Pull request by VincenzoCarlino**
# Description

Added command: `url join`.
Closes: #7725 

# User-Facing Changes

```
Converts a record to url

Search terms: scheme, username, password, hostname, port, path, query, fragment

Usage:
  > url join

Flags:
  -h, --help - Display the help message for this command

Signatures:
  <record> | url join -> <string>

Examples:
  Outputs a url representing the contents of this record
  > {
          "scheme": "http",
          "username": "",
          "password": "",
          "host": "www.pixiv.net",
          "port": "",
          "path": "/member_illust.php",
          "query": "mode=medium&illust_id=99260204",
          "fragment": "",
          "params":
          {
            "mode": "medium",
            "illust_id": "99260204"
          }
        } | url join

  Outputs a url representing the contents of this record
  > {
        "scheme": "http",
        "username": "user",
        "password": "pwd",
        "host": "www.pixiv.net",
        "port": "1234",
        "query": "test=a",
        "fragment": ""
      } | url join

  Outputs a url representing the contents of this record
  > {
        "scheme": "http",
        "host": "www.pixiv.net",
        "port": "1234",
        "path": "user",
        "fragment": "frag"
      } | url join
```                  

# Questions
- Which parameters should be required? Currenlty are: `scheme` and `host`.

### CODE REVIEW

**Comment by fdncred** [comment]
It would be nice to change the amount of spacing to the left of all these strings. It looks kind of weird with `help url join`. 
image

I wonder if it looks better to have multiple entries on one row or if that's more confusing?
```
{
  "scheme": "http", "username": "", "password": "",
  "host": "www.pixiv.net", "port": "",
  "path": "/member_illust.php",
  "query": "mode=medium&illust_id=99260204",
  "fragment": "", "params": {
    "mode": "medium",
    "illust_id": "99260204"
  }
} | url join
```

**Comment by VincenzoCarlino** [comment]
Yes got it for fix the amount of blank spaces.
IMHO multiple entries on one row could be a bit confusing, but it would be a nice hack to reduce the space taken by displaying the help of this command.

**Comment by VincenzoCarlino** [comment]
@fdncred I've "solved" this by adjsuting blank spaces format in string literal of examples. Idk if there is a better way to handle this, I searched on other commands with multiline examples and also in there we play with blank spaces.

Result:
image

**Comment by fdncred** [comment]
It looks better. Thanks.

**Review by fdncred — APPROVED** [review]
I really like this. I wonder if we need to handle input from a `table` as well as a `record` but I'm fine with landing now and seeing if people really need that.

I especially like the good amount of tests. Thanks!

File: crates/nu-command/src/network/url/join.rs [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+                            if let Some(q) = self.query {
+                                if q != s {
+                                    // if query is present it means that also params_span is setted.
+                                    return Err(ShellError::IncompatibleParameters {
+                                        left_message: format!("Mismatch, query param is: {}", s),
+                                        left_span: value.expect_span(),
+                                        right_message: format!("instead qs from params is: {}", q),
+                                        right_span: self
+                                            .params_span
+                                            .expect("params span should be present"),
```
Comment (sholderbach) [inline crates/nu-command/src/network/url/join.rs:260]: I am not immediately convinced that this `.expect` will never panic based on the types going in (albeit sane on the first glance based on the logic). If you can prove this to be safe you could remove the wrapping `Option` or rely on either `Span::unknown` or our `Spanned<T>` wrapper type to link the span to its corresponding value in `UrlComponents`.

File: crates/nu-command/src/network/url/join.rs [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+            scheme_result?,
+            user_and_pwd,
+            host_result?,
+            self.port
+                .map(|p| format!(":{}", p))
+                .as_ref()
+                .unwrap_or(&String::from("")),
+            self.path.as_ref().unwrap_or(&String::from("")),
+            self.query.as_ref().unwrap_or(&String::from("")),
+            self.fragment.as_ref().unwrap_or(&String::from(""))
```
Comment (sholderbach) [inline crates/nu-command/src/network/url/join.rs:323]: Here there is no need to write `String::from("")` and immediately reference it. `Option` allows you to indirectly `Deref` a `String` to `&str`

```suggestion
                .as_deref()
                .unwrap_or_default(),
            self.path.as_deref().unwrap_or_default(),
            self.query.as_deref().unwrap_or_default(),
            self.fragment.as_deref().unwrap_or_default()
```

File: crates/nu-command/src/network/url/join.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+                    if let Some(q) = self.query {
+                        if q != qs {
+                            // if query is present it means that also query_span is setted.
+                            return Err(ShellError::IncompatibleParameters {
+                                left_message: format!("Mismatch, qs from params is: {}", qs),
+                                left_span: value.expect_span(),
+                                right_message: format!("instead query is: {}", q),
+                                right_span: self
+                                    .query_span
+                                    .expect("query param span should be present"),
```
Comment (sholderbach) [inline crates/nu-command/src/network/url/join.rs:200]: Same as above, the linking of value and `Span` is implied.

**Review by sholderbach — COMMENTED** [review]
Thanks for picking up this issue and providing a complete command.
Also great examples and thorough tests!

The only little concern I have if there is a chance that a panic might be introduced (at a later date) in the error handling code paths.

**Comment by sholderbach** [comment]
Thanks for the super quick turnaround!

**Pull request merged**